src/main/java/com/google/devtools/build/lib/rules/genrule/GenRuleBase.java

```diff
@@ -77,6 +80,35 @@ protected RuleConfiguredTargetBuilder updateBuilder(
     return builder;
   }
 
+  enum CommandType {
+    BASH,
+    BATCH,
+    POWERSHELL,
+  }
+
+  private static Pair<CommandType, String> determineCommandTypeAndAttribute(RuleContext ruleContext) {
+    AttributeMap attributeMap = ruleContext.attributes();
+    // TODO(pcloudy): This should match the execution platform instead of using OS.getCurrent()
+    if (OS.getCurrent() == OS.WINDOWS) {
+      if (attributeMap.isAttributeValueExplicitlySpecified("cmd_ps")) {
```

Comment:
No need to check if it's explicitly defined. It's enough to check if it's empty or not. An empty command cannot possibly make sense (it cannot generate any outputs, so the action would fail anyway), so we can ignore `cmd_ps` if the user deliberately sets it to `""`. 

Same for other attributes.

Reply:
That was exactly what I did. But I broke some tests because `cmd` is no longer a mandatory attribute. So I also want to ensure at least one cmd attribute got specified. If I don't use `isAttributeValueExplicitlySpecified`, I cannot tell the value is empty is because the attribute is not set or it's deliberately set to "". And as you said, setting an cmd attribute to "" doesn't make sense, I prefer to not silently ignore it.

Reply:
1. If `cmd` is no longer a mandatory attr, isn't that an incompatible change? Is it mandatory to define at least one of cmd, cmd_bash, cmd_bat, and cmd_ps?

2. Setting the attribute to "" (or None) is customary inside macros:

    ```
    def my_genrule(name, cmd, cmd_bat=None, cmd_ps=None, **kwargs):
      native.genrule(name = name, cmd = cmd, cmd_bat=cmd_bat, cmd_ps=cmd_ps, **kwargs)
    ```

    These attributes are explicitly defined and may be legitimately empty.

Reply:
1. It's mandatory to define at least one of cmd, cmd_bash on Linux and at least one of cmd, cmd_bash, cmd_bat, and cmd_ps on Windows.
For current users that only use `cmd` in genrule, I don't think it is an incompatible change because it will still fail when `cmd` is not set explicitly.

2. I tested, if `cmd_ps = None`, `isAttributeValueExplicitlySpecified("cmd_ps")` is `False`, so it will be ignored. That means if users want to override the value in the macro, they should set the default value to `None` instead of `""`.

Reply:
Thanks!